### PR TITLE
microbatch: set semaphore based on workers

### PR DESCRIPTION
**Pull request by hrmthw**
(Thanks for sending a pull request! Please make sure to read the contribution guidelines, then fill out the blanks below.)

What changes were proposed in this pull request?
------------------------------------------------

Does this close any currently open issues?
------------------------------------------

How was this patch tested?
--------------------------

### CODE REVIEW

File: bentoml/cli/__init__.py [hunk 1]
```diff
@@ -346,6 +346,7 @@ def serve_gunicorn(
                     port=port,
                     num_of_workers=microbatch_workers,
                     prometheus_lock=prometheus_lock,
+                    model_server_workers=workers or 3,
```
Comment (parano) [inline bentoml/cli/__init__.py]: 3 is actually not the default value, I think it should use this function instead to get the default `workers` value:
https://github.com/bentoml/BentoML/blob/master/bentoml/server/utils.py#L29
https://github.com/bentoml/BentoML/blob/master/bentoml/server/gunicorn_config.py#L7

Comment (hrmthw) [reply bentoml/cli/__init__.py]: @parano updated

File: bentoml/marshal/marshal.py [hunk 2]
```diff
@@ -122,10 +122,18 @@ class MarshalService:
...
+    def __init__(
+        self,
+        bento_bundle_path,
+        target_host="localhost",
+        target_port=None,
+        target_count=2,
+    ):
         self.target_host = target_host
         self.target_port = target_port
+        self.target_count = target_count
```
Comment (parano) [inline bentoml/marshal/marshal.py]: add a quick inline comments about `self.target_count` and `self._target_sema`?

File: bentoml/marshal/dispatcher.py [hunk 3]
```diff
@@ -55,17 +54,19 @@ def __init__(self, interval, max_size, task_concurrency=2):
...
     @property
-    @lru_cache(maxsize=1)
     def outbound_sema(self):
         '''
-        semaphore should be created after forked(in the current event loop)
+        semaphore should be created after process forked
         '''
-        return asyncio.Semaphore(self.task_concurrency)
+        if self._sema is None:
+            self._sema = self.shared_sema and self.shared_sema() or asyncio.Semaphore(1)
```
Comment (parano) [inline bentoml/marshal/dispatcher.py]: consider this for readability?
```
self._sema = self.shared_sema() if self.shared_sema else asyncio.Semaphore(1)
```

File: bentoml/cli/__init__.py [hunk 4]
```diff
@@ -346,6 +346,7 @@ def serve_gunicorn(
                     port=port,
                     num_of_workers=microbatch_workers,
                     prometheus_lock=prometheus_lock,
+                    model_server_workers=workers,
```
Comment (parano) [inline bentoml/cli/__init__.py]: is it possible to start GunicornBentoServer before  GunicornMarshalServer? that way you should be able to get the actual worker count via `gunicorn_app.cfg.workers`

Comment (hrmthw) [reply bentoml/cli/__init__.py]: `gunicorn_app.run` will block itself.

File: bentoml/marshal/dispatcher.py (hunk 3 above)
Comment (hrmthw) [reply bentoml/marshal/dispatcher.py]: The later is always better than the former. I always forget it.

File: bentoml/cli/__init__.py (hunk 4 above)
Comment (parano) [reply bentoml/cli/__init__.py]: Ohh got it

Comment (parano) [reply bentoml/cli/__init__.py]: I think a slightly better way to do this is:
```
model_server_workder_count = workers or get_gunicorn_num_of_workers()
```
and then pass it to both GunicornBentoServer and GunicornMarshalServer

That also fixes a current issue, for example when user set the CLI arg `--workers=5`, it still prints something like "get_gunicorn_num_of_workers: 3, calculated by cpu count", which is wrong and very confusing

Comment (hrmthw) [reply bentoml/cli/__init__.py]: Oh yeah, great!

File: bentoml/cli/__init__.py [hunk 5]
```diff
@@ -333,19 +335,23 @@ def serve_gunicorn(
...
             with reserve_free_port() as api_server_port:
                 marshal_server = GunicornMarshalServer(
-                    target_host="localhost",
-                    target_port=api_server_port,
                     bundle_path=bento_service_bundle_path,
                     port=port,
-                    num_of_workers=microbatch_workers,
+                    workers=microbatch_workers,
                     prometheus_lock=prometheus_lock,
+                    outbound_host="localhost",
```
Comment (parano) [inline bentoml/cli/__init__.py]: 👍

File: bentoml/server/gunicorn_config.py [hunk 6]
```diff
@@ -1,11 +1,8 @@
 import atexit
 from multiprocessing.util import _exit_function
 
-from bentoml.server.utils import get_gunicorn_num_of_workers
 from bentoml.utils.usage_stats import track_server_stop
 
-workers = get_gunicorn_num_of_workers()
```
Comment (parano) [inline bentoml/server/gunicorn_config.py]: 👏

**Comment by parano** [comment]
LGTM, thanks for the update @hrmthw, merging now!

**Pull request merged**